cluster-agent/pkg/api/v2/join.go

```diff
@@ -0,0 +1,199 @@
+	}
+	if err := util.RemoveClusterToken(req.ClusterToken); err != nil {
+		return nil, fmt.Errorf("failed to remove cluster token: %w", err)
+	}
+	if !util.HasDqliteLock() {
+		return nil, fmt.Errorf("failed to join the cluster: this is not an HA MicroK8s cluster")
+	}
+
+	// Sanity check cluster agent ports.
+	clusterAgentBind := util.GetServiceArgument("cluster-agent", "--bind")
+	_, port, _ := net.SplitHostPort(clusterAgentBind)
+	if port != req.ClusterAgentPort {
+		return nil, fmt.Errorf("cluster agent port needs to be set to %s", port)
+	}
+
+	// Sanity check node is not in cluster already.
+	hostname := util.GetRemoteHost(req.RemoteHostName, req.RemoteAddress)
+	dqliteCluster, err := util.GetDqliteCluster()
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve dqlite cluster nodes: %w", err)
+	}
+	for _, node := range dqliteCluster {
+		if strings.HasPrefix(node.Address, hostname+":") {
+			return nil, fmt.Errorf("joining node %q is already known to dqlite", hostname)
+		}
+	}
+
+	// Update dqlite cluster if needed
+	if len(dqliteCluster) == 1 && strings.HasPrefix(dqliteCluster[0].Address, "127.0.0.1:") {
+		requestHost, _, _ := net.SplitHostPort(req.HostPort)
+		if err := util.UpdateDqliteIP(ctx, requestHost); err != nil {
+			return nil, fmt.Errorf("failed to update dqlite address to %q: %w", requestHost, err)
+		}
+
+		// Wait for dqlite cluster to come up with new address
+		dqliteCluster, err = util.WaitForDqliteCluster(ctx, func(c util.DqliteCluster) (bool, error) {
+			return len(c) >= 1 && !strings.HasPrefix(c[0].Address, "127.0.0.1:"), nil
+		})
+		if err != nil {
+			return nil, fmt.Errorf("failed waiting for dqlite cluster to come up: %w", err)
+		}
+	}
+
+	callbackToken, err := util.GetOrCreateSelfCallbackToken()
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve self callback token: %w", err)
+	}
+
+	ca, err := util.ReadFile(util.SnapDataPath("certs", "ca.crt"))
+	if err != nil {
+		return nil, fmt.Errorf("failed reading cluster CA: %w", err)
+	}
+	kubeletArgs, err := util.ReadFile(util.SnapDataPath("args", "kubelet"))
+	if err != nil {
+		return nil, fmt.Errorf("failed to read arguments of kubelet service: %w", err)
+	}
+	if hostname != req.RemoteHostName {
+		kubeletArgs = fmt.Sprintf("%s\n--hostname-override=%s", kubeletArgs, hostname)
+	}
+
+	if err := util.MaybePatchCalicoAutoDetectionMethod(ctx, hostname, true); err != nil {
+		return nil, fmt.Errorf("failed to update cni configuration: %w", err)
+	}
+
+	if err := util.CreateNoCertsReissueLock(); err != nil {
+		return nil, fmt.Errorf("failed to create lock file to disable certificate reissuing: %w", err)
+	}
+	response := &JoinResponse{
+		CertificateAuthority:       ca,
+		CallbackToken:              callbackToken,
+		APIServerPort:              util.GetServiceArgument("kube-apiserver", "--secure-port"),
+		APIServerAuthorizationMode: util.GetServiceArgument("kube-apiserver", "--authorization-mode"),
+		HostNameOverride:           hostname,
```

Comment:
In the python implementation the HostNameOverride was returning the IP. Using the hostname is failing for me because it makes it to the dqlite config.